### ISSUE TITLE
`wasm-mutate`: Add/Edit/Remove operations for each Wasm section

### ISSUE THREAD

**Issue by fitzgen** [bytecodealliance/wasm-tools#415]
We should have an add, edit, reorder, and remove mutator for each Wasm section (where it makes sense). This is a tracking issue for each of these things.

In general, removing an entity should be something done optimistically, but which can fail if the entity is referenced in other sections (e.g. if we are trying to remove a global, but a function uses it via `global.get` or `global.set`).

### Custom Sections

* [ ] Add new custom section
* [ ] Edit existing custom section
  * [ ] add bytes
  * [ ] remove bytes
  * [ ] edit bytes
* [ ] Remove custom section
* [ ] Move existing custom section

### Type Section

* [ ] Add new type
* [ ] Edit type (if it isn't used)
* [ ] Remove type
* [ ] Reorder types

### Import Section

* [ ] Add new import
* [ ] Edit import name/module
* [ ] Edit import kind (if it isn't used)
* [ ] Remove import
* [ ] Reorder imports

### Function Section

* [ ] Add new function
* [ ] Add a parameter (and update the code section as necessary)
* [ ] Remove a parameter (and update the code section as necessary)
* [ ] Remove function
* [ ] Reorder functions

### Table Section

* [ ] Add a new table (if there aren't any tables or bulk memory is enabled)
* [ ] Edit an existing table
* [ ] Remove a table
* [ ] Reorder tables

### Memory Section

* [ ] Add a new memory (if there aren't any memories or multi-memory is enabled)
* [ ] Edit an existing memory
* [ ] Remove a memory
* [ ] Reorder memories

### Global Section

* [ ] Add a new global
* [ ] Edit an existing global
* [ ] Remove a global
* [ ] Reorder globals

### Export Section

* [ ] Add a new export
* [X] Rename an existing export
* [ ] Edit an existing export to export a different entity
* [X] Remove an export

### Start Section

* [ ] Add a start section (if there isn't one, and there is a function of the right type)
* [ ] Change which function is the start function in an existing start section (if there is another of the right type)
* [ ] Remove the start section

### Element Section

* [ ] Add a new element segment
* [ ] Edit an existing element segment
  * [ ] Reorder functions in the segment
  * [ ] Add a new function to the segment
  * [ ] Remove a function from the segment
* [ ] Remove an element segment
* [ ] Reorder element segments

### Code Section

I think our peephole and code motion mutators cover the code section pretty well, at least for the level of granularity that this issue is tracking. I'm sure there is stuff we could improve, but these things can be their own issues.

### Data Section

* [ ] Add a new data segment
* [ ] Edit an existing data segment
  * [ ] Reorder bytes within the segment
  * [ ] Add bytes to the segment
  * [ ] Remove bytes from the segment
* [ ] Remove a data segment
* [ ] Reorder data segments

### Data Count Section

Not really applicable. More of something we just have to keep up to date based on mutations of the data section.

**Comment by alexcrichton** [comment]
Most of these removals (and additions) require modifications to the index space which in general can require a whole rewrite of the wasm module, so I think some of those may at least be nontrivial. I've got a branch I started awhile back which implements removing almost all the items here and I just need to finish up the giant `match` statement to rewrite the code section. I hope to push that up in the near future.

**Comment by fitzgen** [comment]
Yes, I was hoping that we could develop a generic function to rewrite the code section with the given renamings and then reuse that function throughout most of these mutators.

**Comment by Jacarte** [comment]
Hi all and happy new year !

I have a doubt on the `ItemRemoveMutator` and I did not find a better way to write it down :(. 

Is there a reason to think that every item removal, excepting for `Data` or `Element` is semantically equivalent ?

https://github.com/bytecodealliance/wasm-tools/blob/6b8da0ec8a283fe04a8a4d012fedc1abe526bac1/crates/wasm-mutate/src/mutators/remove_item.rs#L37 

Thanks in advance

**Comment by fitzgen** [comment]
Happy new year! :)

If the other entities are used (e.g. a function is called or exported) then the mutator will return an error. If they are not used, on the other hand, then removing them still preserves semantics.

**Comment by nagisa** [comment]
> Add a parameter (and update the code section as necessary)
> Remove a parameter (and update the code section as necessary)

This sort of transformation may be made easier if it is preceded by duplicating the type definitions so that each operation and item that refers to a type refers to an unique type definition.

**Comment by nagisa** [comment]
Another transformation that I think could be interesting and which I'm not sure is covered by the issue text here is import-ifying and local-ifying items such as globals and functions. That is replacing something like `(import "" "" (global i32))` with `(global i32 (i32.const 0))` and vice-versa.

Implementing both transformations could also pose a risk of making it difficult to reach a fixpoint too, though.

**Comment by fitzgen** [comment]
> Implementing both transformations could also pose a risk of making it difficult to reach a fixpoint too, though.

We only care about fixpoints when reducing, and in that situation we can avoid the problem by only taking mutations that actually reduce the size of the resulting Wasm binary.

**Comment by fitzgen** [comment]
(Added a bunch more code-related mutation ideas, in case anyone is following along at home, and is interested in taking a look.)
